### PR TITLE
nanopc-t6: add board support for NanoPC-T6

### PR DESCRIPTION
**Pull request by edtubbs**
Based on ROCK 4 support.

Adds nanopc-t6.mk, post-build.sh, U-Boot/Linux config fragments, one Linux dtsi patch, and one U-Boot header patch.

Build commands:
  $ make toolchains
  $ make -j$(nproc)

The bootable image (128 MB) is out/nanopc-t6.img, writable to micro SD.

Tested on NanoPC-T6.

Link: https://wiki.friendlyelec.com/wiki/index.php/NanoPC-T6
Signed-off-by: Ed Tubbs <ectubbs@gmail.com>

### CODE REVIEW

File: br-ext/board/nanopc-t6/post-build.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#! /bin/bash
+# SPDX-License-Identifier: BSD-2-Clause
+
+# Update SSH configuration to allow root login and empty passwords
+sed -i -e 's/#PermitEmptyPasswords no/PermitEmptyPasswords yes/' "$TARGET_DIR"/etc/ssh/sshd_config
+sed -i -e 's/#PermitRootLogin prohibit-password/PermitRootLogin yes/' "$TARGET_DIR"/etc/ssh/sshd_config
+
+# Configure enP2p33s0 with a static IP address
```
Comment (jenswi-linaro) [inline br-ext/board/nanopc-t6/post-build.sh:8]: Doesn't it make more sense with DHCP?

File: nanopc-t6.mk [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+clean: buildroot-clean
+
+include toolchain.mk
+
+################################################################################
+# Arm Trusted Firmware-A
+################################################################################
+TF_A_EXPORTS ?= CROSS_COMPILE="$(CCACHE)$(AARCH64_CROSS_COMPILE)" \
+		M0_CROSS_COMPILE="$(CCACHE)$(AARCH32_CROSS_COMPILE)"
```
Comment (jenswi-linaro) [inline nanopc-t6.mk:71]: The latest version of TF-A requires `CC` and `LD` variables like:
https://github.com/OP-TEE/build/blob/009f86a824c6b1653a22d422696c452c77ab9f85/qemu_v8.mk#L149-L150

I don't know how that affects the `M0_CROSS_COMPILE`.

File: nanopc-t6.mk [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+			 $(ROOT)/build/kconfigs/u-boot_nanopc-t6.conf
+
+UBOOT_FLAGS ?= CROSS_COMPILE=$(CROSS_COMPILE_NS_KERNEL) \
+	       CC=$(CROSS_COMPILE_NS_KERNEL)gcc \
+	       HOSTCC="$(CCACHE) gcc"
+
+$(TPL_BIN):
+	mkdir -p $(BINARIES_PATH)
+	cd $(BINARIES_PATH) && \
+		wget -O $(notdir $(TPL_BIN)) https://github.com/rockchip-linux/rkbin/raw/master/bin/rk35/$(notdir $(TPL_BIN))
```
Comment (jenswi-linaro) [inline nanopc-t6.mk:108]: Doesn't it work with `wget -O $(TPM_BIN) ...` without the `cd $(BINARIES_PATH)`?

File: nanopc-t6.mk [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+.PHONY: u-boot-proper
+u-boot-proper: $(TPL_BIN) $(UBOOT_PATH)/.config u-boot-config
+	$(UBOOT_EXPORTS) $(MAKE) -C $(UBOOT_PATH) $(UBOOT_FLAGS)
+
+.PHONY: u-boot-apply-patch
+u-boot-apply-patch:
+	@if [ -f "$(BOOT_CMD_PATCH)" ]; then \
+		echo "Checking if BOOT_CMD_PATCH has already been applied to $(UBOOT_HEADER)..."; \
+		if patch --dry-run --silent $(UBOOT_HEADER) < $(BOOT_CMD_PATCH) > /dev/null 2>&1; then \
+			patch $(UBOOT_HEADER) < $(BOOT_CMD_PATCH); \
```
Comment (jenswi-linaro) [inline nanopc-t6.mk:136]: This is nasty. Is it necessary?

File: nanopc-t6.mk [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+LINUX_COMMON_FLAGS += ARCH=arm64 CFLAGS_KERNEL="-Wno-error"
+LINUX_COMMON_TARGETS += Image rockchip/rk3588-nanopi6-rev01.dtb \
+			$(if $(filter y,$(LINUX_MODULES)),modules)
+
+.PHONY: linux-apply-patch
+linux-apply-patch:
+	@if [ -f "$(BOARD_DTSI_PATCH)" ]; then \
+		echo "Checking if BOARD_DTSI_PATCH has already been applied to $(LINUX_DTSI)..."; \
+		if patch --dry-run --silent $(LINUX_DTSI) < $(BOARD_DTSI_PATCH) > /dev/null 2>&1; then \
+			patch $(LINUX_DTSI) < $(BOARD_DTSI_PATCH); \
```
Comment (jenswi-linaro) [inline nanopc-t6.mk:177]: This is nasty. Is it necessary?

**Comment by edtubbs** [comment]
Updated based on your feedback, @jenswi-linaro.

File: nanopc-t6.mk [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+UBOOT_DEFCONFIG_FILES := $(UBOOT_PATH)/configs/nanopi6_defconfig \
+			 $(ROOT)/build/kconfigs/u-boot_nanopc-t6.conf
+
+UBOOT_FLAGS ?= CROSS_COMPILE=$(CROSS_COMPILE_NS_KERNEL) \
+	       CC=$(CROSS_COMPILE_NS_KERNEL)gcc \
+	       HOSTCC="$(CCACHE) gcc"
+
+$(TPL_BIN):
+	mkdir -p $(BINARIES_PATH)
+	wget -O $(BINARIES_PATH)/$(notdir $(TPL_BIN)) https://github.com/rockchip-linux/rkbin/raw/master/bin/rk35/$(notdir $(TPL_BIN))
```
Comment (jenswi-linaro) [inline nanopc-t6.mk:108]: TPL_BIN is initialized as `$(BINARIES_PATH)/rk3588_ddr_lp4_2112MHz_lp5_2400MHz_v1.18.bin`. Can't it be used as is for the `-O` argument?

**Review by jenswi-linaro — COMMENTED** [review]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Pull request merged**